services/121-service/src/notifications/whatsapp/whatsapp.service.ts

```diff
@@ -73,51 +74,74 @@ export class WhatsappService {
       payload['messageContentType'] = messageContentType;
     }
 
-    let errorOccurred = false;
-    let messageToStore;
     try {
-      messageToStore = await twilioClient.messages.create(payload);
+      const messageToStore = await twilioClient.messages.create(payload);
+      await this.storeSendWhatsapp({
+        message: messageToStore,
+        userId,
+        registrationId,
+        mediaUrl,
+        messageContentType,
+        messageProcessType,
+      });
       return messageToStore.sid;
     } catch (error) {
-      errorOccurred = true;
-      messageToStore = {
-        accountSid: process.env.TWILIO_SID,
-        body: payload.body,
-        mediaUrl,
-        to: payload.to,
-        messagingServiceSid: process.env.TWILIO_MESSAGING_SID,
-        dateCreated: new Date().toISOString(),
-        sid: `failed-${uuid()}`,
-        status: 'failed',
-        errorCode: error.code,
-      };
-      throw error;
-    } finally {
-      messageToStore.userId = userId;
-      await this.storeSendWhatsapp(
-        messageToStore,
+      await this.storeSendWhatsapp({
+        message: {
+          accountSid: process.env.TWILIO_SID ?? '',
+          body: payload.body ?? '',
+          to: payload.to,
+          messagingServiceSid: process.env.TWILIO_MESSAGING_SID ?? '',
+          dateCreated: new Date(),
+          sid: `failed-${uuid()}`,
+          status: 'failed',
+          errorCode: error.code,
+          errorMessage: error.message,
+        },
+        userId,
         registrationId,
         mediaUrl,
         messageContentType,
         messageProcessType,
-        errorOccurred ? undefined : existingSidToUpdate,
-      );
+        existingSidToUpdateDueToFailure: existingSidToUpdate,
+      });
+      throw error;
     }
   }
 
-  public async storeSendWhatsapp(
+  public async storeSendWhatsapp({
     message,
-    registrationId: number | undefined,
-    mediaUrl: string | null | undefined,
-    messageContentType?: MessageContentType,
-    messageProcessType?: MessageProcessType,
-    existingSidToUpdate?: string,
-  ): Promise<void> {
+    userId,
+    registrationId,
+    mediaUrl,
+    messageContentType,
+    messageProcessType,
+    existingSidToUpdateDueToFailure,
+  }: {
+    message: Pick<
```

Comment:
Should we consider changing this to a class as we use it in 2 places?

Reply:
I considered it, but then I considered that there is a potential scenario where the two functions deviate, and they need different properties from `MessageInstance`. If that isn't likely, then I think a better refactor would be to reuse the same function altogether, rather than reuse the same type. I also considered doing that, but wasn't sure where to put the function so that both services would have access to it.